### PR TITLE
Do not warn about an oblivious mismatch for an explicitly implemented interface.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #35619.

Only the last commit needs a review, previous are coming from https://github.com/dotnet/roslyn/pull/35613.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Symbols/Source/ExplicitInterfaceHelpers.cs [hunk 1]
```diff
@@ -207,7 +207,7 @@ public static string GetMemberNameWithoutInterfaceName(string fullName)
             // interface in its base class list that contains a member ..."
             MultiDictionary<NamedTypeSymbol, NamedTypeSymbol>.ValueSet set = containingType.InterfacesAndTheirBaseInterfacesNoUseSiteDiagnostics[explicitInterfaceNamedType];
             int setCount = set.Count;
-            if (setCount == 0 || !set.Contains(explicitInterfaceNamedType))
+            if (setCount == 0 || !set.Contains(explicitInterfaceNamedType, TypeSymbol.EqualsObliviousNullableModifierMatchesAny))
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/ExplicitInterfaceHelpers.cs]: >set.Contains(explicitInterfaceNamedType, TypeSymbol.EqualsObliviousNullableModifierMatchesAny) [](start = 34, length = 94)

Is this a linear search? How long might this list be? Would it make sense to check `set.Contains(explicitInterfaceNamedType)` first?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/ExplicitInterfaceHelpers.cs]: Haven't looked very closely but it seems to delegate to ImmutableHashSet, which will call GetHashCode on the TypeSymbolComparer, which is probably related to why this work depended on fixing GetHashCode for constructed types with annotated type arguments.

---
In reply to: [297426710](https://github.com/dotnet/roslyn/pull/36761#discussion_r297426710) [](ancestors = 297426710)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/ExplicitInterfaceHelpers.cs]: > Is this a linear search? How long might this list be? Would it make sense to check set.Contains(explicitInterfaceNamedType) first?

Yes, it is a linear search among the duplicates that differ only by tuple names and nullability. I don't think a lot of duplicates will be common, often there will be only one type in the set. 


---
In reply to: [297428012](https://github.com/dotnet/roslyn/pull/36761#discussion_r297428012) [](ancestors = 297428012,297426710)

**Pull request merged**